### PR TITLE
Assignment5+6

### PR DESCRIPTION
**Pull request by schuywalker**
@Schuywalker Schuyler Asplin and @A-Tanner Ray Tanner. Monday Submission.

### CODE REVIEW

**Comment by MarkCapka** [comment]
Working on code review for this Pull request - Mark Capka

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,30 +1,105 @@
...
 {
     public class SampleData : ISampleData
     {
         // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        public IEnumerable<string> CsvRows // = the current return ? 2/17 lecture
+        {
+            get
+            {
+                return File.ReadAllLines(FilePath).Skip(1).ToList();//System.Reflection.Assembly.GetExecutingAssembly().Location; current will work but it relies on unsafe assumptions
```
Comment (xiokd) [inline Assignment/Assignment/SampleData.cs:16]: A possible approach when using System.Reflection.Assembly.GetExecutingAssembly().Location, would be to use it with the System.IO.GetDirectoryName method to get the absolute path of the executing assembly's directory returned as a string and then append the relative path to the file to that string. This approach however will need to be checked for null.

**Review by xiokd — COMMENTED** [review]
Overall everything was good. There was one possible change that could be suggested.

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of Spokane-based addresses. ✔
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
   - Be sure that `Person.Address` is also populated. ✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ✔

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use the `people` parameter from `ISampleData.GetUniqueListOfStates` for your data source. ✔
   - At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.  

## Extra Credit

- Implement the homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `IAsyncSampleData`). Refactor your `SampleData` and `SampleDataAsync` classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌✔

## Fundamentals

- Ensure you enable:
  - nullable reference types is enabled ✔
  - net6 targeted ✔
  - C# 10.0 ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, favor using Assert.AreEqual<T>() (the generic version) ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

File: Assignment/Assignment/SampleData.cs (hunk 1 above)
Comment (A-Tanner) [reply Assignment/Assignment/SampleData.cs:16]: Thanks for the feedback, that answers a question I've been having about how to extract a directory from the commented method (which returns the path to a .dll file)

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,105 @@
...
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+
+            // Define Query
+            IEnumerable<string> stateQuery = (
+                from line in CsvRows
+                let columnSeparatedLines = line.Split(',')
+                orderby columnSeparatedLines[6]
+                select columnSeparatedLines[6]
+                ).Distinct().ToList();
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:42]: I would suggest sticking with the LINQ method syntax, such as below.
```suggestion
            return CsvRows
                .Select(row => row.Split(',')[6])
                .OrderBy(state => state)
                .Distinct();
```

**Review by mmwoodfo — COMMENTED** [review]
Good job!

**Comment by mmwoodfo** [comment]
@schuywalker did you do a PR review?

**Pull request closed without merging**